### ISSUE TITLE
Bing Maps Deprecation

### ISSUE THREAD

**Issue by hl662** [iTwin/itwinjs-core#7009]
**Is your feature request related to a problem? Please describe.**
On June 30, Microsoft [disabled the ability](https://www.microsoft.com/en-us/maps/bing-maps/discontinued-services) for users to generate new Bing Map keys. Existing hyperlinks that pointed our users to create a bing Map key would be redirected to the Url above, recommending migrating to Azure Maps. Existing Bing Map keys will continue to be supported until June 20, 2028, after which the service will be shut down (as explained in Microsoft's deprecation notice).

I'll highlight two components in core that make use of Bing Maps, and the effect of the deprecation:

- [BingLocationProvider](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/BingLocation.ts): Azure Maps has the [same feature available](https://learn.microsoft.com/en-us/azure/azure-maps/migrate-find-location-query), with some slight differences in the response body of the new endpoint. We can create a new AzureLocationProvider to support the same functionality.
- [BingElevationProvider](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/tile/map/BingElevation.ts): Azure Maps [deprecated](https://azure.microsoft.com/en-us/updates/azure-maps-elevation-apis-and-render-v2-dem-tiles-will-be-retired-on-5-may-2023/) their elevation API as of May last year, and there doesn't seem to be an easy way to replace that. This might warrant more discussion on alternatives.

**Describe the solution you'd like**
- In 4.8 or 4.9, depending on how quick the PR moves, we will flag relevant APIs and components that make use of BIngMaps as deprecated, without modifying the existing default behavior.

- We need to reenable the AzureMapsLayerFormat, revisit core's ability to support and use AzureMaps, without modifying the existing behavior of fetching Bing Maps for now. In 5.x, we can look to migrate to another map tile imagery provider, we do have 4 years before our hand is forced...

**Comment by aruniverse** [comment]
**Reviving this issue with updated status and a concrete plan.**

### Version context

4.x reaches end-of-life on **June 13, 2026** (~7 weeks away), so no further work can land there. All changes must target **5.x** (currently at 5.9.0-dev).

### Deprecation timeline under the support policy

Per the [API deprecation policy](https://github.com/iTwin/itwinjs-core/blob/master/docs/learning/api-support-policies.md#api-deprecation-policy), `@public` APIs require a **1-year grace period** after being marked `@deprecated` before they can be removed in any following major release. If we deprecate in 5.x today, removal is possible in **6.0 at the earliest, after ~April 2027**.

Bing Maps for Enterprise fully shuts down **June 30, 2028**. This gives us a workable window: deprecate in 5.x now → remove in 6.x before June 2028.

### Current state (April 2026)

None of the planned 4.8/4.9 deprecation work landed. Auditing `master` today, no `@deprecated` tags exist on any of the affected APIs. The three areas of impact are:

**1. `BingLocationProvider` (`@public`)** — geocoding via `dev.virtualearth.net/REST/v1/Locations`. Used directly inside `ViewGlobeLocationTool`, making the built-in "fly to location" tool silently broken for any app without a working key.

- **Replacement**: Azure Maps has a direct equivalent — the [Geocoding API](https://learn.microsoft.com/en-us/azure/azure-maps/migrate-find-location-query) (`atlas.microsoft.com/geocode`). A new `AzureLocationProvider` can wrap this with the same `getLocation(query)` signature and would require an Azure Maps key supplied at startup alongside the existing `cesiumIonKey` pattern.

**2. `BingMaps` imagery + `"BingProvider"` default (`@public`)** — `BackgroundMapProviderName` defaults to `"BingProvider"` and this is persisted in iModel view definitions. `BingMapsImageryLayerProvider` hits `dev.virtualearth.net/REST/v1/Imagery/Metadata/*`.

- **Replacement**: `AzureMapsMapLayerFormat` and `AzureMapsLayerImageryProvider` are already implemented in MapLayerImageryFormats.ts but commented out from the registered formats list. Re-enabling this is the path forward. Changing the default `BackgroundMapProviderName` away from `"BingProvider"` is a breaking change that must wait for 6.0 with full migration guidance.

**3. `BingElevationProvider` (`@public`)** — three point elevation endpoints at `dev.virtualearth.net/REST/v1/Elevation/*`. Used internally by `IModelConnection.computeProjectElevationRange` and `queryTerrainElevationOffset` when navigating to global locations.

- **Replacement**: **There is no direct drop-in.** Azure Maps [deprecated their elevation API in May 2023](https://azure.microsoft.com/en-us/updates/azure-maps-elevation-apis-and-render-v2-dem-tiles-will-be-retired-on-5-may-2023/). Cesium Ion only provides terrain as quantized-mesh tiles — there is no REST endpoint to query elevation at a point. The most viable built-in fallback is the existing `ApproximateTerrainHeights` class, which provides baked-in elevation ranges from CGIAR/USGS data. It won't match Bing's precision but is already used as a pre-fetch heuristic in `CesiumTerrainProvider`. The longer-term fix is to make elevation lookups pluggable (allow apps to supply their own provider) so consumers can integrate Open-Elevation, AWS Terrain Tiles, or other services. This needs a design discussion.

### Proposed action items

**5.x (deprecation, additive — no breaking changes):**
- [ ] Add `@deprecated in 5.x.0 - will not be removed until after YYYY-MM-DD` tags to `BingLocationProvider`, `BingElevationProvider`, and `ImageryMapLayerFormatId = "BingMaps"` / `BackgroundMapProviderName = "BingProvider"` pointing to their replacements
- [ ] Re-enable `AzureMapsMapLayerFormat` and add an `AzureLocationProvider` as the replacements
- [ ] Update docs and MapLayerSources.json to stop surfacing Bing entries as defaults
- [ ] Have `BingElevationProvider`'s internal callers (`IModelConnection`, `ViewGlobalLocation`) fall back to `ApproximateTerrainHeights` when no Bing key is configured

**6.x (breaking):**
- [ ] Change the default `BackgroundMapProviderName` away from `"BingProvider"` with migration guidance
- [ ] Remove `BingLocationProvider`, `BingElevationProvider`, `"BingMaps"` format, and the `BingProvider` string literal from the public API surface
- [ ] Finalize the pluggable elevation provider design so consumer apps can supply their own

The elevation provider story needs the most discussion — happy to open a separate issue for that design if that would help move this forward.

**Comment by aruniverse** [comment]
> 2. BingMaps imagery + "BingProvider" default (@public) — BackgroundMapProviderName defaults to "BingProvider" and this is persisted in iModel view definitions. 
>
> **On "persisted in iModel view definitions" — what this means and the migration risks**
>
> The `"BingProvider"` string is baked into the iModel's element data, not just runtime settings.

1. A `DisplayStyle` element stored in the iModel's SQLite database contains a `jsonProperties.styles` blob.
2. Inside that blob, `mapImagery.backgroundBase.provider.name` holds the string `"BingProvider"`, along with a `url` containing `https://dev.virtualearth.net/...`.
3. The older `backgroundMap.providerName` field (deprecated in 3.6, kept for backward compatibility in `PersistentBackgroundMapProps`) can also contain `"BingProvider"`, and `BackgroundMapSettings.toPersistentJSON()` still writes it for non-default providers.
4. `BackgroundMapProvider.fromJSON()` hard-codes the fallback: any unrecognized `name` value silently falls back to `"BingProvider"` (`props.name === "MapBoxProvider" ? props.name : "BingProvider"`).

**Concrete repercussions:**

- **Existing iModels will not automatically migrate.** Any iModel saved today with the default background map has `"BingProvider"` + a `dev.virtualearth.net` URL written into its element table. When Bing keys stop working in June 2028, those views will silently show no background map unless the iModel is explicitly re-saved with updated settings or the application overrides the base layer at runtime.

- **The `fromJSON` fallback means "BingProvider" is invisible.** Because any unknown `name` coerces to `"BingProvider"`, removing the string from the type union in 6.x won't cause a runtime failure for old iModel data — it will just silently continue attempting Bing. This is actually a migration hazard: apps may think they've migrated when they haven't.

- **Display style overrides and saved views compound this.** `DisplayStyleSettings.toOverrides({ includeBackgroundMap: true })` serializes the provider name into override JSON that can be stored outside the iModel (in databases, configuration files, or iTwin Platform saved views). All of those stored blobs independently need migration — there is no single "update the iModel" fix.

- **The `providerName` / `providerData` legacy fields in the old `BackgroundMapProps` format are also persisted.** `toPersistentJSON()` writes them for backward compatibility. These fields explicitly contain the string `"BingProvider"`. Even if the new `mapImagery` path is migrated, old reader code that reads the legacy fields won't know the difference.

**What safe migration looks like:**

The safest approach is for the application (not the library) to inspect the `backgroundBase.provider.name` at view load time and replace it with an `AzureProvider` equivalent if it is `"BingProvider"`. The library can facilitate this by:
- Providing a utility function (e.g., `DisplayStyleSettings.migrateProviders(mapping)`) that rewrites known deprecated provider names
- Ensuring that after 6.x removes `"BingProvider"` from `BackgroundMapProviderName`, `fromJSON` either throws on an unknown name or returns a clear `undefined`/neutral default rather than silently re-instating Bing

The `dev.virtualearth.net` URL stored in `mapImagery.backgroundBase.url` also needs to be replaced — that URL is what the `BingMapsImageryLayerProvider` actually fetches the tile metadata from. Simply changing the provider name is not enough; the `url` field must be updated to an Azure Maps equivalent simultaneously.

**Comment by hl662** [comment]
Quick status update after #9300:

We now have a decent imagery story: Azure Maps basemap support was added and lives in `@itwin/map-layers-formats`, 
 instead of adding more provider-specific code to `core-frontend`.

The remaining Bing-dependent areas are location and elevation.

- **Location** looks replaceable. `BingLocationProvider` is still constructed directly in `core-frontend`, but Azure Maps geocoding seems close enough to the current contract that this should be solvable (with me bot). One question would just be where to 'house' the implementation. map-layers-formats seems wrong, the APIs in there are for map imagery. So unless its okay to put the azure location implementation in core-frontend, we may need a new map-services pkg? @@
- **Elevation** is the hard part. `BingElevationProvider` is still used by `IModelConnection`, `queryTerrainElevationOffset()`, and `MapTileTree`, so core itself depends on Bing-backed answers today. Azure Maps does not provide a replacement here, though this comment has defined valid approaches, involving a new pluggable elevation provider interface (this can go into core-frontend). Then the implementation can lie outside of core-frontend.


Elevation has two different things: terrain height, and geoid offset lookup.
One useful clue is the existing [TODO](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/tile/map/CesiumTerrainProvider.ts#L129) in `CesiumTerrainProvider`: we already load terrain tiles for rendering, but we do not yet expose an API to sample heights from that data. Seems like an avenue for abstracting an interface for terrain height lookup, and then have CesiumTerrainProvider implement that interface. geoid? Not sure...

Seems to me like we would expose core-owned hook points, interfaces for:
- location lookup
- terrain height lookup
- geoid offset lookup

Then concrete implementations like Bing and Azure can live outside `core-frontend`, similar to the imagery work.
However, this would take more time and effort obviously, and I'll be deprecating the bing imagery stuff first. Location and Elevation, I am unsure about deprecating those, not until we have some solutions or at least the interfaces set up in core frontend, so we can at least inform consumers they would have to supply their own implementation (if we can't do one ourselves right now)

**Comment by hl662** [comment]
Final update: After https://github.com/iTwin/itwinjs-core/pull/9378 and https://github.com/iTwin/itwinjs-core/pull/9412, we have exposed new interfaces that deprecated Imagery, Location and Elevation providers that are implemented through Bing, and added generic interfaces that are plug and play into core-frontend. 

With this, downstream apps can provide their own implementations for the 3 big features above that aligns with what our tiling and tools expect.

Closing this issue.
